<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="clustree">
<title>Plotting clustering trees • clustree</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plotting clustering trees">
<meta property="og:description" content="clustree">
<meta property="og:image" content="https://lazappi.github.io/clustree/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google tag (gtag.js) --><script async src="https://www.googletagmanager.com/gtag/js?id=G-6GT3P9FP6F"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-6GT3P9FP6F');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">clustree</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.0.9001</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/clustree.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/lazappi/clustree/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Plotting clustering trees</h1>
                        <h4 data-toc-skip class="author">Luke
Zappia</h4>
            
            <h4 data-toc-skip class="date">Last updated: 12 June
2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lazappi/clustree/blob/HEAD/vignettes/clustree.Rmd" class="external-link"><code>vignettes/clustree.Rmd</code></a></small>
      <div class="d-none name"><code>clustree.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="what-is-a-clustering-tree">What is a clustering tree?<a class="anchor" aria-label="anchor" href="#what-is-a-clustering-tree"></a>
</h2>
<p>Clustering analysis is used in many contexts to group similar
samples. One problem when conducting this kind of analysis is how many
clusters to use. This is usually controlled by a parameter provided to
the clustering algorithm, such as <span class="math inline">\(k\)</span>
for <span class="math inline">\(k\)</span>-means clustering.</p>
<p>Statistics designed to help you make this choice typically either
compare two clusterings or score a single clustering. A clustering tree
is different in that it visualises the relationships between at a range
of resolutions.</p>
<p>To build a clustering tree we need to look at how cells move as the
clustering resolution is increased. Each cluster forms a node in the
tree and edges are constructed by considering the cells in a cluster at
a lower resolution (say <span class="math inline">\(k = 2\)</span>) that
end up in a cluster at the next highest resolution (say <span class="math inline">\(k = 3\)</span>). By connecting clusters in this
way we can see how clusters are related to each other, which are clearly
distinct and which are unstable. Extra information about the cells in
each node can also be overlaid in order to help make the decision about
which resolution to use. For more information about clustering trees
please refer to our associated publication <span class="citation">(Zappia and Oshlack 2018)</span>.</p>
</div>
<div class="section level2">
<h2 id="a-simple-example">A simple example<a class="anchor" aria-label="anchor" href="#a-simple-example"></a>
</h2>
<p>To demonstrate what a clustering tree looks like we will work through
a short example using the <code>nba_clusts</code> dataset.</p>
<div class="section level3">
<h3 id="the-data">The data<a class="anchor" aria-label="anchor" href="#the-data"></a>
</h3>
<p>The <code>nba_clusts</code> dataset consists of some basic statistics
from 150 NBA players in 2017, 50 from each of three positions (Center,
Point Guard and Shooting Guards). This dataset has then been clustered
using <span class="math inline">\(k\)</span>-means with different values
of <span class="math inline">\(k\)</span>. For more information see
<code><a href="../reference/nba_clusts.html">?nba_clusts</a></code>. Let’s load the data and take a look:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lazappi/clustree" class="external-link">clustree</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ggraph</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"nba_clusts"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">nba_clusts</span><span class="op">)</span></span>
<span><span class="co">#&gt;   Position TurnoverPct ReboundPct AssistPct FieldGoalPct K1 K2 K3 K4 K5</span></span>
<span><span class="co">#&gt; 1   Center        10.6       17.4      14.7         50.2  1  2  2  1  2</span></span>
<span><span class="co">#&gt; 2   Center         9.6       19.6      16.9         46.8  1  2  2  1  2</span></span>
<span><span class="co">#&gt; 3   Center        16.0       14.2       5.4         57.1  1  2  2  1  2</span></span>
<span><span class="co">#&gt; 4   Center        17.2       14.9       2.7         58.3  1  2  2  1  2</span></span>
<span><span class="co">#&gt; 5   Center        11.7       20.8       4.4         55.7  1  2  2  1  2</span></span>
<span><span class="co">#&gt; 6   Center        12.5       17.4       7.9         54.5  1  2  2  1  2</span></span>
<span><span class="co">#&gt;          PC1       PC2</span></span>
<span><span class="co">#&gt; 1  -5.907675 -4.559588</span></span>
<span><span class="co">#&gt; 2  -3.398585 -4.015772</span></span>
<span><span class="co">#&gt; 3 -14.705825 -3.907384</span></span>
<span><span class="co">#&gt; 4 -17.489016 -3.598523</span></span>
<span><span class="co">#&gt; 5 -17.705077 -3.455650</span></span>
<span><span class="co">#&gt; 6 -13.069032 -3.845442</span></span></code></pre></div>
<p>Here we have a <code>data.frame</code> with player positions and some
statistics (turnover percentage, rebound percentage, assist percentage
and field goal percentage) as well as some additional columns. These
columns contain the cluster assignments from clustering this data using
<span class="math inline">\(k\)</span>-means with values of <span class="math inline">\(k\)</span> from <span class="math inline">\(k =
1\)</span> to <span class="math inline">\(k = 5\)</span>.</p>
</div>
<div class="section level3">
<h3 id="plotting-a-tree">Plotting a tree<a class="anchor" aria-label="anchor" href="#plotting-a-tree"></a>
</h3>
<p>This clustering information is all we need to build a clustering
tree. Each column must consist of numeric values indicating which
cluster each sample has been assigned to. To plot the tree we just pass
this information to the <code>clustree</code> function. We also need to
specify a <code>prefix</code> string to indicate which columns contain
the clusterings.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-plot-1.png" width="700"></p>
<p>We can see that one cluster is very distinct and does not change with
the value of <span class="math inline">\(k\)</span>. This is the Center
players which are very different to the other positions. On the other
side of the tree we see a single cluster that splits into the two
clusters we would expect to see. After this the tree becomes messier and
there are node with multiple incoming edges. This is a good indication
that we have over clustered the data.</p>
</div>
<div class="section level3">
<h3 id="controlling-aesthetics">Controlling aesthetics<a class="anchor" aria-label="anchor" href="#controlling-aesthetics"></a>
</h3>
<p>By default the size of each node is related to the number of samples
in each cluster and the colour indicates the clustering resolution.
Edges are coloured according to the number of samples they represent and
the transparency shows the incoming node proportion, the number of
samples in the edge divided by the number of samples in the node it
points to. We can control these aesthetics by setting them to specific
values:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, node_colour <span class="op">=</span> <span class="st">"purple"</span>, node_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>         node_alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-aes-static-1.png" width="700"></p>
<p>We can also link these aesthetics to other information we have about
the samples. All the additional columns in the dataset are available to
be added as attributes to the nodes in our tree. Because each node
represents multiple samples we need to supply an aggregation function to
use as well specifying a column name. Let’s try colouring the nodes
according to the rebound percentage:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, node_colour <span class="op">=</span> <span class="st">"ReboundPct"</span>,</span>
<span>         node_colour_aggr <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-aes-1.png" width="700"></p>
<p>We can clearly see that the distinct cluster containing the Center
players has a higher rebound percentage on average compared to the other
clusters.</p>
<div class="section level4">
<h4 id="sc3-stability-index">SC3 stability index<a class="anchor" aria-label="anchor" href="#sc3-stability-index"></a>
</h4>
<p>Apart from information in the dataset itself it can useful to display
measures of clustering quality as aesthetics. The stability index from
the <strong>{SC3}</strong> package <span class="citation">(Kiselev et
al. 2017)</span> measures the stability of clusters across resolutions
and is automatically calculated when a clustering tree is built. It can
be accessed by setting an aesthetic to <code>"sc3_stability"</code> and
because it is calculated by cluster we don’t need to provide an
aggregation function. For example:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, node_colour <span class="op">=</span> <span class="st">"sc3_stability"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-stability-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="layout">Layout<a class="anchor" aria-label="anchor" href="#layout"></a>
</h3>
<p>By default the tree is drawn using the Reingold-Tilford tree layout
algorithm which tries to place nodes below their parents <span class="citation">(Reingold and Tilford 1981)</span>. Alternatively we
could use the Sugiyama layout by specifying the <code>layout</code>
argument. This algorithm tries to minimise the number of crossing edges
<span class="citation">(Sugiyama, Tagawa, and Toda 1981)</span> and can
produce more attractive trees in some cases.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, layout <span class="op">=</span> <span class="st">"sugiyama"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-layout-1.png" width="700"></p>
<p>For both of these layout algorithms <code><a href="../reference/clustree.html">clustree()</a></code> uses
slightly modified versions of the tree by default. Only the core network
of edges, those that are the highest in-proportion edge for a node, are
used when creating the layout. In most cases this leads to more
attractive trees that are easier to interpret. To turn this off, and use
all edges for deciding the layout, we can set
<code>use_core_edges</code> to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, layout <span class="op">=</span> <span class="st">"sugiyama"</span>, use_core_edges <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-layout-nocore-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="adding-labels">Adding labels<a class="anchor" aria-label="anchor" href="#adding-labels"></a>
</h3>
<p>To make it easy to identify clusters the cluster nodes are labelled
with their cluster number (controlled using the <code>node_text</code>
arguments) but sometimes it is useful to add labels with additional
information. This is done the same way as the other aesthetics. Here we
label nodes with the maximum assist percentage:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, node_label <span class="op">=</span> <span class="st">"AssistPct"</span>,</span>
<span>         node_label_aggr <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-labels-1.png" width="700"></p>
<p>One way this can be useful is if we have assigned labels to the
samples. Here is a custom function that labels a cluster if all the
players are the same position, otherwise it labels the cluster as
“mixed”:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">label_position</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">labels</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">labels</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">position</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">labels</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">position</span> <span class="op">&lt;-</span> <span class="st">"mixed"</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">position</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, node_label <span class="op">=</span> <span class="st">"Position"</span>,</span>
<span>         node_label_aggr <span class="op">=</span> <span class="st">"label_position"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-labels-custom-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="clustering-trees-for-scrna-seq-data">Clustering trees for scRNA-seq data<a class="anchor" aria-label="anchor" href="#clustering-trees-for-scrna-seq-data"></a>
</h2>
<p>Clustering has become a core tool for analysing single-cell
RNA-sequencing (scRNA-seq) datasets. These datasets contain gene
expression measurements from hundreds to hundreds of thousands of cells.
Often samples come from complex tissues containing many types of cells
and clustering is used to group similar cells together. To make it
easier to produce clustering trees for these kinds of datasets we
provide interfaces for some of the objects commonly used to analyse
scRNA-seq data.</p>
<p>The clustree package contains an example simulated scRNA-seq data
that has been clustered using the <strong>{SC3}</strong> and
<strong>{Seurat}</strong> <span class="citation">(Satija et al.
2015)</span> packages.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sc_example"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sc_example</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "counts"          "logcounts"       "tsne"            "sc3_clusters"   </span></span>
<span><span class="co">#&gt; [5] "seurat_clusters"</span></span></code></pre></div>
<div class="section level3">
<h3 id="singlecellexperiment-objects">
<code>SingleCellExperiment</code> objects<a class="anchor" aria-label="anchor" href="#singlecellexperiment-objects"></a>
</h3>
<pre><code><span><span class="co">#&gt; Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by</span></span>
<span><span class="co">#&gt; 'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</span></span></code></pre>
<p>The <code>SingleCellExperiment</code> is one of these common objects,
used across a range of Bioconductor packages. Let’s have a look at an
example, but first we need to convert the example dataset to a
<code>SingleCellExperiment</code> object:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">sc_example</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>                                          logcounts <span class="op">=</span> <span class="va">sc_example</span><span class="op">$</span><span class="va">logcounts</span><span class="op">)</span>,</span>
<span>                            colData <span class="op">=</span> <span class="va">sc_example</span><span class="op">$</span><span class="va">sc3_clusters</span>,</span>
<span>                            reducedDims <span class="op">=</span> <span class="fu">SimpleList</span><span class="op">(</span>TSNE <span class="op">=</span> <span class="va">sc_example</span><span class="op">$</span><span class="va">tsne</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The clustering information is held in the <code>coldata</code>
slot.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 8 columns</span></span>
<span><span class="co">#&gt;       sc3_1_clusters sc3_2_clusters sc3_3_clusters sc3_4_clusters</span></span>
<span><span class="co">#&gt;             &lt;factor&gt;       &lt;factor&gt;       &lt;factor&gt;       &lt;factor&gt;</span></span>
<span><span class="co">#&gt; Cell1              1              2              3              4</span></span>
<span><span class="co">#&gt; Cell2              1              2              3              2</span></span>
<span><span class="co">#&gt; Cell3              1              1              1              1</span></span>
<span><span class="co">#&gt; Cell4              1              1              1              3</span></span>
<span><span class="co">#&gt; Cell5              1              2              3              4</span></span>
<span><span class="co">#&gt; Cell6              1              2              2              2</span></span>
<span><span class="co">#&gt;       sc3_5_clusters sc3_6_clusters sc3_7_clusters sc3_8_clusters</span></span>
<span><span class="co">#&gt;             &lt;factor&gt;       &lt;factor&gt;       &lt;factor&gt;       &lt;factor&gt;</span></span>
<span><span class="co">#&gt; Cell1              3              2              2              3</span></span>
<span><span class="co">#&gt; Cell2              2              1              1              1</span></span>
<span><span class="co">#&gt; Cell3              5              4              4              5</span></span>
<span><span class="co">#&gt; Cell4              4              3              3              4</span></span>
<span><span class="co">#&gt; Cell5              3              2              5              8</span></span>
<span><span class="co">#&gt; Cell6              2              1              1              1</span></span></code></pre></div>
<p>We can plot a clustering tree in the same way we did with a
<code>data.frame</code>. In this case the clustering column names
contain a suffix that needs to be stripped away, so we will pass that
along as well.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">sce</span>, prefix <span class="op">=</span> <span class="st">"sc3_"</span>, suffix <span class="op">=</span> <span class="st">"_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/sce-plot-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="seurat-objects">
<code>Seurat</code> objects<a class="anchor" aria-label="anchor" href="#seurat-objects"></a>
</h3>
<blockquote>
<p><strong>NOTE:</strong> This section requires the Seurat package
(&gt;= 5.0.0). This package isn’t installed so the results won’t be
shown.</p>
</blockquote>
<p>Clustering trees can also be produced directly from
<code>Seurat</code> objects. Let’s convert our
<code>SingleCellExperiment</code> to <code>Seurat</code> format:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://satijalab.org/seurat" class="external-link">"Seurat"</a></span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create the Seurat object, Seurat &gt;= expects a sparse matrix</span></span>
<span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">sc_example</span><span class="op">$</span><span class="va">counts</span>, <span class="st">"sparseMatrix"</span><span class="op">)</span>,</span>
<span>                             data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">sc_example</span><span class="op">$</span><span class="va">logcounts</span>, <span class="st">"sparseMatrix"</span><span class="op">)</span>,</span>
<span>                             meta.data <span class="op">=</span> <span class="va">sc_example</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add the t-SNE embedding</span></span>
<span><span class="va">seurat</span><span class="op">[[</span><span class="st">'TSNE'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateDimReducObject.html" class="external-link">CreateDimReducObject</a></span><span class="op">(</span>embeddings <span class="op">=</span> <span class="va">sc_example</span><span class="op">$</span><span class="va">tsne</span>,</span>
<span>                                         key <span class="op">=</span> <span class="st">"tSNE_"</span><span class="op">)</span></span></code></pre></div>
<p>In this case the clustering information is held in the
<code>meta.data</code> slot which can be accessed using the
<code>[[</code> operator:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>We can now produce a clustering tree using this object. In this
example the prefix for clustering columns is <code>res.</code> but in
most cases the default prefix from <code>Seurat</code> will be
automatically used.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">seurat</span>, prefix <span class="op">=</span> <span class="st">"res."</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Note: This example uses the newer Seurat object available in version
3.0.0 or greater of Seurat. There is also an interface for the older
seurat object but this may be deprecated in the future.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="using-genes-as-aesthetics">Using genes as aesthetics<a class="anchor" aria-label="anchor" href="#using-genes-as-aesthetics"></a>
</h3>
<p>As well as being able to use any additional columns for aesthetics we
can also use the expression of individual genes. Let’s colour the nodes
in the <code>Seurat</code> tree by <code>Gene730</code> (a highly
variable gene). Again we need to supply an aggregation function.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">seurat</span>, prefix <span class="op">=</span> <span class="st">"res."</span>,</span>
<span>         node_colour <span class="op">=</span> <span class="st">"Gene730"</span>, node_colour_aggr <span class="op">=</span> <span class="st">"median"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="overlaying-clustering-trees">Overlaying clustering trees<a class="anchor" aria-label="anchor" href="#overlaying-clustering-trees"></a>
</h2>
<p>One way to think of clustering trees is that they add an extra
dimension to the data that shows how clusters “evolve” over time
(increasing resolution). Some times it can be useful to overlay this
dimension on other dimensions in the data, particularly those that come
from dimensionality reduction techniques. We can do this using the
<code>clustree_overlay</code> function:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree_overlay.html">clustree_overlay</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, x_value <span class="op">=</span> <span class="st">"PC1"</span>, y_value <span class="op">=</span> <span class="st">"PC2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-overlay-1.png" width="700"></p>
<p>The easiest way to understand this plot is to imagine that you are
looking down on the clustering tree from above. The x and y axes are the
chosen dimensions in the data and the z axis is the clustering
resolution. Each cluster node is placed at the mean x and y values of
the samples it contains. We can also see points corresponding to the
individual samples.</p>
<div class="section level3">
<h3 id="choosing-what-to-colour">Choosing what to colour<a class="anchor" aria-label="anchor" href="#choosing-what-to-colour"></a>
</h3>
<p>Due to the way <code>ggplot2</code> works we can only colour one
element in the plot (the node points actually use the fill aesthetic).
By default the tree edges are coloured according to the clustering
resolution they originate from. Alternatively we can choose to colour
the sample points by their highest resolution cluster. The colour of
whichever element isn’t using the colour aesthetic can be set using the
<code>alt_colour</code> argument.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree_overlay.html">clustree_overlay</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, x_value <span class="op">=</span> <span class="st">"PC1"</span>, y_value <span class="op">=</span> <span class="st">"PC2"</span>,</span>
<span>                 use_colour <span class="op">=</span> <span class="st">"points"</span>, alt_colour <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-overlay-colour-1.png" width="700"></p>
<p>Aesthetics for the clustering tree nodes can be controlled in the
same way as for the regular <code>clustree</code> function and there are
additional arguments for adjusting the appearance of edges and
points.</p>
</div>
<div class="section level3">
<h3 id="labelling-nodes">Labelling nodes<a class="anchor" aria-label="anchor" href="#labelling-nodes"></a>
</h3>
<p>One of the downsides of this visualisation is that it can be hard to
identify the clusters that each node represents as they can often
overlap. To make this a bit easier we can label each node with the
resolution and cluster ID.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree_overlay.html">clustree_overlay</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, x_value <span class="op">=</span> <span class="st">"PC1"</span>, y_value <span class="op">=</span> <span class="st">"PC2"</span>,</span>
<span>                 label_nodes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-overlay-labels-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="showing-the-side-view">Showing the side view<a class="anchor" aria-label="anchor" href="#showing-the-side-view"></a>
</h3>
<p>While the main overlay plot show us the tree from above it can also
be useful to see it from the side, showing one of the x or y dimensions
against the resolution dimension. We can get these views by setting the
<code>plot_sides</code> option to <code>TRUE</code>. This will return a
list of <code>ggplot</code> objects instead of a single plot.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">overlay_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustree_overlay.html">clustree_overlay</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span>, x_value <span class="op">=</span> <span class="st">"PC1"</span>,</span>
<span>                                 y_value <span class="op">=</span> <span class="st">"PC2"</span>, plot_sides <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">overlay_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "overlay" "x_side"  "y_side"</span></span>
<span></span>
<span><span class="va">overlay_list</span><span class="op">$</span><span class="va">x_side</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-overlay-sides-1.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">overlay_list</span><span class="op">$</span><span class="va">y_side</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/nba-overlay-sides-2.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="modifying-appearance">Modifying appearance<a class="anchor" aria-label="anchor" href="#modifying-appearance"></a>
</h2>
<p>The <code><a href="../reference/clustree.html">clustree()</a></code> function returns a <code>ggplot</code>
object which can be modified using functions in the
<strong>{ggplot2}</strong> or <strong>{ggraph}</strong> packages. For
example we could change the colour scales used for the nodes and
edges:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_brewer</span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html" class="external-link">scale_edge_color_continuous</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"blue"</span>, high <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">edge_colour</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">edge_colour</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="clustree_files/figure-html/modify-1.png" width="700"></p>
<div class="section level3">
<h3 id="legends">Legends<a class="anchor" aria-label="anchor" href="#legends"></a>
</h3>
<p>The way <code>ggplot</code> objects is displayed is relative to the
size of the plotting window or output file. While the main plot will
always fit sometimes legends will be placed outside the visible area.
One solution to this is to simply increase the size of the image. An
alternative solution is to turn off some of the legends, either by
setting some of the aesthetics to static values or by using the
<code>guides</code> function. We could also move them to somewhere they
might fit better. For example let’s remove the edge legends and move the
rest to the bottom:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">nba_clusts</span>, prefix <span class="op">=</span> <span class="st">"K"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">guides</span><span class="op">(</span>edge_colour <span class="op">=</span> <span class="cn">FALSE</span>, edge_alpha <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `&lt;scale&gt;` argument of `guides()` cannot be `FALSE`. Use "none" instead as</span></span>
<span><span class="co">#&gt; of ggplot2 3.3.4.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p><img src="clustree_files/figure-html/legends-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="citing-clustree">Citing clustree<a class="anchor" aria-label="anchor" href="#citing-clustree"></a>
</h2>
<p>If you find <strong>{clustree}</strong> or the clustering trees
approach useful for your work please cite our associated
publication:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"clustree"</span><span class="op">)</span></span>
<span><span class="co">#&gt; To cite package 'clustree' in publications use:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   Zappia L, Oshlack A. Clustering trees: a visualization for evaluating</span></span>
<span><span class="co">#&gt;   clusterings at multiple resolutions. Gigascience. 2018;7.</span></span>
<span><span class="co">#&gt;   DOI:gigascience/giy083</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   @Article{,</span></span>
<span><span class="co">#&gt;     author = {Luke Zappia and Alicia Oshlack},</span></span>
<span><span class="co">#&gt;     title = {Clustering trees: a visualization for evaluating clusterings at</span></span>
<span><span class="co">#&gt;              multiple resolutions},</span></span>
<span><span class="co">#&gt;     journal = {GigaScience},</span></span>
<span><span class="co">#&gt;     volume = {7},</span></span>
<span><span class="co">#&gt;     number = {7},</span></span>
<span><span class="co">#&gt;     month = {jul},</span></span>
<span><span class="co">#&gt;     year = {2018},</span></span>
<span><span class="co">#&gt;     url = {https://doi.org/10.1093/gigascience/giy083},</span></span>
<span><span class="co">#&gt;     doi = {10.1093/gigascience/giy083},</span></span>
<span><span class="co">#&gt;   }</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Kiselev2017-an" class="csl-entry">
Kiselev, Vladimir Yu, Kristina Kirschner, Michael T Schaub, Tallulah
Andrews, Andrew Yiu, Tamir Chandra, Kedar N Natarajan, et al. 2017.
<span>“<span class="nocase">SC3: consensus clustering of single-cell
RNA-seq data</span>.”</span> <em>Nature Methods</em> 14 (5): 483–86. <a href="https://doi.org/10.1038/nmeth.4236" class="external-link">https://doi.org/10.1038/nmeth.4236</a>.
</div>
<div id="ref-Reingold1981-iy" class="csl-entry">
Reingold, E M, and J S Tilford. 1981. <span>“<span class="nocase">Tidier
Drawings of Trees</span>.”</span> <em>IEEE Transactions on Software
Engineering</em> SE-7 (2): 223–28. <a href="https://doi.org/10.1109/TSE.1981.234519" class="external-link">https://doi.org/10.1109/TSE.1981.234519</a>.
</div>
<div id="ref-Satija2015-or" class="csl-entry">
Satija, Rahul, Jeffrey A Farrell, David Gennert, Alexander F Schier, and
Aviv Regev. 2015. <span>“<span class="nocase">Spatial reconstruction of
single-cell gene expression data</span>.”</span> <em>Nature
Biotechnology</em> 33 (5): 495–502. <a href="https://doi.org/10.1038/nbt.3192" class="external-link">https://doi.org/10.1038/nbt.3192</a>.
</div>
<div id="ref-Sugiyama1981-wu" class="csl-entry">
Sugiyama, K, S Tagawa, and M Toda. 1981. <span>“<span class="nocase">Methods for Visual Understanding of Hierarchical System
Structures</span>.”</span> <em>IEEE Transactions on Systems, Man, and
Cybernetics</em> 11 (2): 109–25. <a href="https://doi.org/10.1109/TSMC.1981.4308636" class="external-link">https://doi.org/10.1109/TSMC.1981.4308636</a>.
</div>
<div id="ref-Zappia2018-lz" class="csl-entry">
Zappia, Luke, and Alicia Oshlack. 2018. <span>“<span class="nocase">Clustering trees: a visualization for evaluating
clusterings at multiple resolutions</span>.”</span> <em>GigaScience</em>
7 (7). <a href="https://doi.org/10.1093/gigascience/giy083" class="external-link">https://doi.org/10.1093/gigascience/giy083</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Luke Zappia, Alicia Oshlack.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
